kolibri/logger/management/commands/userdata_utils.py

```diff
@@ -0,0 +1,327 @@
+        for i in range(0, n_to_create):
+            Facility.objects.create(name='Test Facility {i}'.format(i=i + 1))
+    return Facility.objects.all()[0:n_facilities]
+
+
+def get_or_create_classrooms(**options):
+    n_classes = options['n_classes']
+    facility = options['facility']
+    n_on_device = Classroom.objects.filter(parent=facility).count()
+    n_to_create = n_classes - n_on_device
+    if n_to_create > 0:
+        print('Generating {n} classroom object(s) for facility: {name}'.format(
+            n=n_to_create,
+            name=facility.name,
+        ))
+        for i in range(0, n_to_create):
+            Classroom.objects.create(
+                parent=facility,
+                name='Classroom {i}{a}'.format(i=i + 1, a=random.choice('ABCD'))
+            )
+    return Classroom.objects.filter(parent=facility)[0:n_classes]
+
+
+def get_or_create_classroom_users(**options):
+    classroom = options['classroom']
+    n_users = options['n_users']
+    user_data = options['user_data']
+    facility = options['facility']
+
+    # The headers in the user_data.csv file that we use to generate user Full Names
+    # Note, we randomly pick from these to give deliberately varied (and sometimes idiosyncratic)
+    # Full names - because we should never assume that users have names like us
+    user_data_name_fields = ["GivenName", "MiddleInitial", "Surname"]
+
+    n_on_device = HierarchyRelationsFilter(FacilityUser.objects.all()).filter_by_hierarchy(
+        target_user=F("id"),
+        ancestor_collection=classroom,
+    ).count()
+
+    # Only generate new users if there are fewer users than requested.
+    n_to_create = n_users - n_on_device
+    if n_to_create > 0:
+        print('Generating {n} user object(s) for class: {classroom} in facility: {facility}'.format(
+            n=n_to_create,
+            classroom=classroom,
+            facility=facility,
+        ))
+        for i in range(0, n_to_create):
+            # Get the first base data that does not have a matching user already
+            base_data = user_data[n_on_device + i]
+            # Randomly create the name from 1 to 3 of the three user name fields
+            name = " ".join([base_data[key] for key in random.sample(user_data_name_fields, random.randint(1, 3))])
+            user = FacilityUser.objects.create(
+                facility=facility,
+                full_name=name,
+                username=base_data['Username']
+            )
+            # Set a dummy password so that if we want to login as this learner later, we can.
+            user.set_password('password')
+            # Add the user to the current classroom
+            classroom.add_member(user)
+
+    return HierarchyRelationsFilter(FacilityUser.objects.all()).filter_by_hierarchy(
+        target_user=F("id"),
+        ancestor_collection=classroom,
+    )[0:n_users]
+
+
+def add_channel_activity_for_user(**options): # noqa: max-complexity=16
```

Comment:
Is the `noqa` flag still necessary here? (possibly, but with your slight refactor, it might have dipped).